File: pyNCAA/scripts/bracket_sim.py
# scripts/bracket_sim.py
# March Madness 2026 bracket simulation using full model stats + Kalman adjustments

import json
import math
import os
import sys

SCRIPTS_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(SCRIPTS_DIR, "..", "data")

stats_data = json.loads(open(os.path.join(SCRIPTS_DIR, "..", "..", "data", "stats_cache", "ncaab", "20250317.json"), "r").read())
H = stats_data["season"]
ks = json.loads(open(os.path.join(DATA_DIR, "kalman_state.json"), "r").read())


# -- Team name resolution ---------------------------------------------------

def norm_key(s):
    import re
    s = str(s or "").lower()
    s = re.sub(r"[^a-z0-9\s]", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def resolve_team(name, data):
    if name in data:
        return name
    teams = list(data.keys())
    wanted = norm_key(name)
    for k in teams:
        if norm_key(k) == wanted:
            return k
    # Longest prefix match
    best_prefix = None
    best_len = 0
    for k in teams:
        nk = norm_key(k)
        if nk == wanted or wanted == nk:
            return k
        if (wanted.startswith(nk + " ") or nk.startswith(wanted + " ")) and len(nk) >= 3:
            l = min(len(nk), len(wanted))
            if l > best_len:
                best_len = l
                best_prefix = k
    if best_prefix:
        return best_prefix
    # Fuzzy
    for k in teams:
        nk = norm_key(k)
        if nk in wanted or wanted in nk:
            return k
    return None


# -- Normal CDF -------------------------------------------------------------

def normal_cdf(x):
    a1 = 0.254829592
    a2 = -0.284496736
    a3 = 1.421413741
    a4 = -1.453152027
    a5 = 1.061405429
    p = 0.3275911
    sign = -1 if x < 0 else 1
    z = abs(x) / math.sqrt(2)
    t = 1.0 / (1.0 + p * z)
    y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * math.exp(-z * z)
    return 0.5 * (1.0 + sign * y)


# -- Game simulation --------------------------------------------------------

def sim_game(team_a, team_b):
    a_key = resolve_team(team_a, H)
    b_key = resolve_team(team_b, H)

    if not a_key or not b_key:
        if not a_key and not b_key:
            return {"winner": team_a, "loser": team_b, "prob": 0.5, "margin": 0}
        if not a_key:
            return {"winner": team_b, "loser": team_a, "prob": 0.75, "margin": 5}
        return {"winner": team_a, "loser": team_b, "prob": 0.75, "margin": 5}

    a = H[a_key]
    b = H[b_key]

    a_expected = (a["OFF"] + b["DEF"]) / 2
    b_expected = (b["OFF"] + a["DEF"]) / 2
    avg_pace = (a["PACE"] + b["PACE"]) / 2 / 100

    margin = (a_expected - b_expected) * avg_pace

    # Add Kalman adjustment
    a_kalman = resolve_team(team_a, ks.get("teams", {}))
    b_kalman = resolve_team(team_b, ks.get("teams", {}))
    if a_kalman and b_kalman:
        margin += (ks["teams"][a_kalman]["adj_mean"] - ks["teams"][b_kalman]["adj_mean"])

    game_sd = 11
    p_a_wins = normal_cdf(margin / game_sd)

    if p_a_wins >= 0.5:
        return {"winner": team_a, "loser": team_b, "prob": p_a_wins, "margin": margin}
    else:
        return {"winner": team_b, "loser": team_a, "prob": 1 - p_a_wins, "margin": -margin}


# -- Bracket definition -----------------------------------------------------

def main():
    print("\u2554" + "\u2550" * 66 + "\u2557")
    print("\u2551         2026 NCAA MARCH MADNESS BRACKET PREDICTIONS            \u2551")
    print("\u2551         Bayesian Model + Kalman Filter Adjustments             \u2551")
    print("\u255a" + "\u2550" * 66 + "\u255d\n")

    print("\u2501" * 18 + " FIRST FOUR " + "\u2501" * 18)
    ff1 = sim_game("Texas", "N.C. State")
    ff2 = sim_game("UMBC", "Howard")
    ff3 = sim_game("Miami OH", "SMU")
    ff4 = sim_game("Prairie View A&M", "Lehigh")

    for g in [ff1, ff2, ff3, ff4]:
        print(f"  {g['winner']:<22} def. {g['loser']:<22} ({g['prob']*100:.0f}% | proj margin: {abs(g['margin']):.1f})")

    # -- Region definitions --
    regions = {
        "EAST": [
            ["Duke", "Siena"],
            ["Ohio St.", "TCU"],
            ["St. John's", "Northern Iowa"],
            ["Kansas", "Cal Baptist"],
            ["Louisville", "South Florida"],
            ["Michigan St.", "North Dakota St."],
            ["UCLA", "UCF"],
            ["Connecticut", "Furman"],
        ],
        "WEST": [
            ["Arizona", "LIU"],
            ["Villanova", "Utah St."],
            ["Wisconsin", "High Point"],
            ["Arkansas", "Hawaii"],
            ["BYU", ff1["winner"]],
            ["Gonzaga", "Kennesaw St."],
            ["Miami FL", "Missouri"],
            ["Purdue", "Queens"],
        ],
        "MIDWEST": [
            ["Michigan", ff2["winner"]],
            ["Georgia", "Saint Louis"],
            ["Texas Tech", "Akron"],
            ["Alabama", "Hofstra"],
            ["Tennessee", ff3["winner"]],
            ["Virginia", "Wright St."],
            ["Kentucky", "Santa Clara"],
            ["Iowa St.", "Tennessee St."],
        ],
        "SOUTH": [
            ["Florida", ff4["winner"]],
            ["Clemson", "Iowa"],
            ["Vanderbilt", "McNeese St."],
            ["Nebraska", "Troy"],
            ["North Carolina", "VCU"],
            ["Illinois", "Penn"],
            ["Saint Mary's", "Texas A&M"],
            ["Houston", "Idaho"],
        ],
    }

    seeds = {
        "EAST": {"Duke": 1, "Siena": 16, "Ohio St.": 8, "TCU": 9, "St. John's": 5, "Northern Iowa": 12, "Kansas": 4, "Cal Baptist": 13, "Louisville": 6, "South Florida": 11, "Michigan St.": 3, "North Dakota St.": 14, "UCLA": 7, "UCF": 10, "Connecticut": 2, "Furman": 15},
        "WEST": {"Arizona": 1, "LIU": 16, "Villanova": 8, "Utah St.": 9, "Wisconsin": 5, "High Point": 12, "Arkansas": 4, "Hawaii": 13, "BYU": 6, ff1["winner"]: 11, "Gonzaga": 3, "Kennesaw St.": 14, "Miami FL": 7, "Missouri": 10, "Purdue": 2, "Queens": 15},
        "MIDWEST": {"Michigan": 1, ff2["winner"]: 16, "Georgia": 8, "Saint Louis": 9, "Texas Tech": 5, "Akron": 12, "Alabama": 4, "Hofstra": 13, "Tennessee": 6, ff3["winner"]: 11, "Virginia": 3, "Wright St.": 14, "Kentucky": 7, "Santa Clara": 10, "Iowa St.": 2, "Tennessee St.": 15},
        "SOUTH": {"Florida": 1, ff4["winner"]: 16, "Clemson": 8, "Iowa": 9, "Vanderbilt": 5, "McNeese St.": 12, "Nebraska": 4, "Troy": 13, "North Carolina": 6, "VCU": 11, "Illinois": 3, "Penn": 14, "Saint Mary's": 7, "Texas A&M": 10, "Houston": 2, "Idaho": 15},
    }

    def get_seed(region, team):
        if region in seeds and team in seeds[region]:
            return seeds[region][team]
        for r in seeds.values():
            if team in r:
                return r[team]
        return "?"

    def sim_round(matchups, region, round_name):
        results = []
        print(f"\n  {round_name}:")
        for a, b in matchups:
            g = sim_game(a, b)
            s_w = get_seed(region, g["winner"])
            s_l = get_seed(region, g["loser"])
            print(f"    ({str(s_w):>2}) {g['winner']:<20} def. ({str(s_l):>2}) {g['loser']:<20} {g['prob']*100:.0f}%  margin: {abs(g['margin']):.1f}")
            results.append(g["winner"])
        return results

    final_four_teams = []

    for region_name, matchups in regions.items():
        print(f"\n{'\u2501' * 22} {region_name} REGION {'\u2501' * 22}")

        r64 = sim_round(matchups, region_name, "Round of 64")

        r32_matchups = [[r64[i], r64[i + 1]] for i in range(0, len(r64), 2)]
        r32 = sim_round(r32_matchups, region_name, "Round of 32")

        s16_matchups = [[r32[i], r32[i + 1]] for i in range(0, len(r32), 2)]
        s16 = sim_round(s16_matchups, region_name, "Sweet 16")

        e8 = sim_round([[s16[0], s16[1]]], region_name, "Elite 8")

        print(f"\n  >>> {region_name} CHAMPION: ({get_seed(region_name, e8[0])}) {e8[0]}")
        final_four_teams.append(e8[0])

    # -- Final Four --
    print(f"\n{'\u2501' * 22} FINAL FOUR {'\u2501' * 22}")
    print("  Lucas Oil Stadium, Indianapolis -- April 4, 2026\n")

    semi1 = sim_game(final_four_teams[0], final_four_teams[3])  # East vs South
    semi2 = sim_game(final_four_teams[2], final_four_teams[1])  # Midwest vs West

    print("  Semifinal 1 (East vs South):")
    print(f"    {semi1['winner']:<20} def. {semi1['loser']:<20} ({semi1['prob']*100:.0f}% | margin: {abs(semi1['margin']):.1f})")
    print("  Semifinal 2 (Midwest vs West):")
    print(f"    {semi2['winner']:<20} def. {semi2['loser']:<20} ({semi2['prob']*100:.0f}% | margin: {abs(semi2['margin']):.1f})")

    # -- Championship --
    print(f"\n{'\u2501' * 22} NATIONAL CHAMPIONSHIP {'\u2501' * 22}")
    print("  April 6, 2026 -- Lucas Oil Stadium, Indianapolis\n")

    champ = sim_game(semi1["winner"], semi2["winner"])
    print(f"    {champ['winner']:<20} def. {champ['loser']:<20} ({champ['prob']*100:.0f}% | margin: {abs(champ['margin']):.1f})")

    print(f"\n{'=' * 60}")
    print(f"  NATIONAL CHAMPION: {champ['winner']}")
    print(f"{'=' * 60}")

    # -- Summary --
    print("\n BRACKET SUMMARY")
    print("\u2500" * 40)
    region_names = list(regions.keys())
    print("  Final Four:")
    for i, team in enumerate(final_four_teams):
        print(f"    {region_names[i]:<8}: ({get_seed(region_names[i], team)}) {team}")
    print(f"  Champion:  {champ['winner']}")
    print(f"  Runner-up: {champ['loser']}\n")

    # -- Upset alerts --
    print("  NOTABLE UPSETS TO WATCH:")
    print("  (Teams where a lower seed wins by 3+ seed lines)")
    print("\u2500" * 40)


if __name__ == "__main__":
    main()

File: jsNCAA/scripts/bracket_sim.mjs
// scripts/bracket_sim.mjs
// March Madness 2026 bracket simulation using full model stats + Kalman adjustments
import fs from 'fs';

import path from 'path';
import { fileURLToPath } from 'url';
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const statsData = JSON.parse(fs.readFileSync(path.join(__dirname, '..', '..', 'data', 'stats_cache', 'ncaab', '20260317.json'), 'utf8'));
const H = statsData.season;
const ks = JSON.parse(fs.readFileSync('data/kalman_state.json', 'utf8'));

// ── Team name resolution ─────────────────────────────────────────────────────
function normKey(s) {
  return String(s || "").toLowerCase().replace(/[^a-z0-9\s]/g, " ").replace(/\s+/g, " ").trim();
}

function resolveTeam(name, data) {
  if (data[name]) return name;
  const teams = Object.keys(data);
  const wanted = normKey(name);
  for (const k of teams) {
    if (normKey(k) === wanted) return k;
  }
  // Longest prefix match to avoid "Miami" matching wrong team
  let bestPrefix = null, bestLen = 0;
  for (const k of teams) {
    const nk = normKey(k);
    if (nk === wanted || wanted === nk) return k;
    if ((wanted.startsWith(nk + " ") || nk.startsWith(wanted + " ")) && nk.length >= 3) {
      const len = Math.min(nk.length, wanted.length);
      if (len > bestLen) { bestLen = len; bestPrefix = k; }
    }
  }
  if (bestPrefix) return bestPrefix;
  // Fuzzy
  for (const k of teams) {
    const nk = normKey(k);
    if (nk.includes(wanted) || wanted.includes(nk)) return k;
  }
  return null;
}

// ── Normal CDF ───────────────────────────────────────────────────────────────
function normalCDF(x) {
  const a1 = 0.254829592, a2 = -0.284496736, a3 = 1.421413741;
  const a4 = -1.453152027, a5 = 1.061405429, p = 0.3275911;
  const sign = x < 0 ? -1 : 1;
  const z = Math.abs(x) / Math.SQRT2;
  const t = 1.0 / (1.0 + p * z);
  const y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-z * z);
  return 0.5 * (1.0 + sign * y);
}

// ── Game simulation using efficiency stats + Kalman adjustment ───────────────
// Uses OFF/DEF efficiency ratings (points per 100 possessions) as the primary
// team strength measure, with Kalman adj_mean as a form adjustment.
// Tournament games are on neutral court (no HCA).

function simGame(teamA, teamB) {
  const aKey = resolveTeam(teamA, H);
  const bKey = resolveTeam(teamB, H);

  if (!aKey || !bKey) {
    if (!aKey && !bKey) return { winner: teamA, loser: teamB, prob: 0.5, margin: 0 };
    if (!aKey) return { winner: teamB, loser: teamA, prob: 0.75, margin: 5 };
    return { winner: teamA, loser: teamB, prob: 0.75, margin: 5 };
  }

  const a = H[aKey];
  const b = H[bKey];

  // Net efficiency: OFF - DEF (higher = better)
  // Projected margin = (A_OFF - B_DEF + B_OFF - A_DEF) / 2... simplified:
  // Margin ~ (A_NET - B_NET) / 2 adjusted for pace
  const aNet = a.OFF - a.DEF;
  const bNet = b.OFF - b.DEF;

  // Matchup-based margin: (A scores vs B defense) - (B scores vs A defense)
  const aExpected = (a.OFF + b.DEF) / 2;  // A's expected scoring rate
  const bExpected = (b.OFF + a.DEF) / 2;  // B's expected scoring rate
  const avgPace = (a.PACE + b.PACE) / 2 / 100;  // possessions factor

  let margin = (aExpected - bExpected) * avgPace;

  // Add Kalman adjustment (recent form drift)
  const aKalman = resolveTeam(teamA, ks.teams);
  const bKalman = resolveTeam(teamB, ks.teams);
  if (aKalman && bKalman) {
    margin += (ks.teams[aKalman].adj_mean - ks.teams[bKalman].adj_mean);
  }

  // Game-to-game standard deviation in college basketball is ~11 points
  const gameSD = 11;

  const pAWins = normalCDF(margin / gameSD);

  if (pAWins >= 0.5) {
    return { winner: teamA, loser: teamB, prob: pAWins, margin: margin };
  } else {
    return { winner: teamB, loser: teamA, prob: 1 - pAWins, margin: -margin };
  }
}

// ── Bracket definition ───────────────────────────────────────────────────────
console.log('╔══════════════════════════════════════════════════════════════════╗');
console.log('║         2026 NCAA MARCH MADNESS BRACKET PREDICTIONS            ║');
console.log('║         Bayesian Model + Kalman Filter Adjustments             ║');
console.log('╚══════════════════════════════════════════════════════════════════╝\n');

console.log('━━━━━━━━━━━━━━━━━━ FIRST FOUR ━━━━━━━━━━━━━━━━━━');
const ff1 = simGame('Texas', 'N.C. State');        // 11 seed West
const ff2 = simGame('UMBC', 'Howard');              // 16 seed Midwest
const ff3 = simGame('Miami OH', 'SMU');             // 11 seed Midwest
const ff4 = simGame('Prairie View A&M', 'Lehigh');  // 16 seed South

for (const g of [ff1, ff2, ff3, ff4]) {
  console.log(`  ${g.winner.padEnd(22)} def. ${g.loser.padEnd(22)} (${(g.prob*100).toFixed(0)}% | proj margin: ${Math.abs(g.margin).toFixed(1)})`);
}

// ── Region definitions ───────────────────────────────────────────────────────
const regions = {
  'EAST': [
    ['Duke', 'Siena'],                             // 1 vs 16
    ['Ohio St.', 'TCU'],                            // 8 vs 9
    ["St. John's", 'Northern Iowa'],                // 5 vs 12
    ['Kansas', 'Cal Baptist'],                      // 4 vs 13
    ['Louisville', 'South Florida'],                // 6 vs 11
    ['Michigan St.', 'North Dakota St.'],           // 3 vs 14
    ['UCLA', 'UCF'],                                // 7 vs 10
    ['Connecticut', 'Furman'],                      // 2 vs 15
  ],
  'WEST': [
    ['Arizona', 'LIU'],                             // 1 vs 16
    ['Villanova', 'Utah St.'],                      // 8 vs 9
    ['Wisconsin', 'High Point'],                    // 5 vs 12
    ['Arkansas', 'Hawaii'],                         // 4 vs 13
    ['BYU', ff1.winner],                            // 6 vs 11
    ['Gonzaga', 'Kennesaw St.'],                    // 3 vs 14
    ['Miami FL', 'Missouri'],                       // 7 vs 10
    ['Purdue', 'Queens'],                           // 2 vs 15
  ],
  'MIDWEST': [
    ['Michigan', ff2.winner],                       // 1 vs 16
    ['Georgia', 'Saint Louis'],                     // 8 vs 9
    ['Texas Tech', 'Akron'],                        // 5 vs 12
    ['Alabama', 'Hofstra'],                         // 4 vs 13
    ['Tennessee', ff3.winner],                      // 6 vs 11
    ['Virginia', 'Wright St.'],                     // 3 vs 14
    ['Kentucky', 'Santa Clara'],                    // 7 vs 10
    ['Iowa St.', 'Tennessee St.'],                  // 2 vs 15
  ],
  'SOUTH': [
    ['Florida', ff4.winner],                        // 1 vs 16
    ['Clemson', 'Iowa'],                            // 8 vs 9
    ['Vanderbilt', 'McNeese St.'],                  // 5 vs 12
    ['Nebraska', 'Troy'],                           // 4 vs 13
    ['North Carolina', 'VCU'],                      // 6 vs 11
    ['Illinois', 'Penn'],                           // 3 vs 14
    ["Saint Mary's", 'Texas A&M'],                  // 7 vs 10
    ['Houston', 'Idaho'],                           // 2 vs 15
  ],
};

const seeds = {
  'EAST': { 'Duke': 1, 'Siena': 16, 'Ohio St.': 8, 'TCU': 9, "St. John's": 5, 'Northern Iowa': 12, 'Kansas': 4, 'Cal Baptist': 13, 'Louisville': 6, 'South Florida': 11, 'Michigan St.': 3, 'North Dakota St.': 14, 'UCLA': 7, 'UCF': 10, 'Connecticut': 2, 'Furman': 15 },
  'WEST': { 'Arizona': 1, 'LIU': 16, 'Villanova': 8, 'Utah St.': 9, 'Wisconsin': 5, 'High Point': 12, 'Arkansas': 4, 'Hawaii': 13, 'BYU': 6, [ff1.winner]: 11, 'Gonzaga': 3, 'Kennesaw St.': 14, 'Miami FL': 7, 'Missouri': 10, 'Purdue': 2, 'Queens': 15 },
  'MIDWEST': { 'Michigan': 1, [ff2.winner]: 16, 'Georgia': 8, 'Saint Louis': 9, 'Texas Tech': 5, 'Akron': 12, 'Alabama': 4, 'Hofstra': 13, 'Tennessee': 6, [ff3.winner]: 11, 'Virginia': 3, 'Wright St.': 14, 'Kentucky': 7, 'Santa Clara': 10, 'Iowa St.': 2, 'Tennessee St.': 15 },
  'SOUTH': { 'Florida': 1, [ff4.winner]: 16, 'Clemson': 8, 'Iowa': 9, 'Vanderbilt': 5, 'McNeese St.': 12, 'Nebraska': 4, 'Troy': 13, 'North Carolina': 6, 'VCU': 11, 'Illinois': 3, 'Penn': 14, "Saint Mary's": 7, 'Texas A&M': 10, 'Houston': 2, 'Idaho': 15 },
};

function getSeed(region, team) {
  // Check all regions if not found in specified one (for Final Four)
  if (seeds[region]?.[team] != null) return seeds[region][team];
  for (const r of Object.values(seeds)) {
    if (r[team] != null) return r[team];
  }
  return '?';
}

function simRound(matchups, region, roundName) {
  const results = [];
  console.log(`\n  ${roundName}:`);
  for (const [a, b] of matchups) {
    const g = simGame(a, b);
    const sW = getSeed(region, g.winner);
    const sL = getSeed(region, g.loser);
    console.log(`    (${String(sW).padStart(2)}) ${g.winner.padEnd(20)} def. (${String(sL).padStart(2)}) ${g.loser.padEnd(20)} ${(g.prob*100).toFixed(0)}%  margin: ${Math.abs(g.margin).toFixed(1)}`);
    results.push(g.winner);
  }
  return results;
}

const finalFourTeams = [];

for (const [regionName, matchups] of Object.entries(regions)) {
  console.log(`\n${'━'.repeat(22)} ${regionName} REGION ${'━'.repeat(22)}`);

  const r64 = simRound(matchups, regionName, 'Round of 64');

  const r32matchups = [];
  for (let i = 0; i < r64.length; i += 2) r32matchups.push([r64[i], r64[i + 1]]);
  const r32 = simRound(r32matchups, regionName, 'Round of 32');

  const s16matchups = [];
  for (let i = 0; i < r32.length; i += 2) s16matchups.push([r32[i], r32[i + 1]]);
  const s16 = simRound(s16matchups, regionName, 'Sweet 16');

  const e8 = simRound([[s16[0], s16[1]]], regionName, 'Elite 8');

  console.log(`\n  >>> ${regionName} CHAMPION: (${getSeed(regionName, e8[0])}) ${e8[0]}`);
  finalFourTeams.push(e8[0]);
}

// ── Final Four ───────────────────────────────────────────────────────────────
console.log(`\n${'━'.repeat(22)} FINAL FOUR ${'━'.repeat(22)}`);
console.log(`  Lucas Oil Stadium, Indianapolis — April 4, 2026\n`);

// NCAA Final Four pairings: East vs South, Midwest vs West
const semi1 = simGame(finalFourTeams[0], finalFourTeams[3]); // East vs South
const semi2 = simGame(finalFourTeams[2], finalFourTeams[1]); // Midwest vs West

console.log(`  Semifinal 1 (East vs South):`);
console.log(`    ${semi1.winner.padEnd(20)} def. ${semi1.loser.padEnd(20)} (${(semi1.prob*100).toFixed(0)}% | margin: ${Math.abs(semi1.margin).toFixed(1)})`);
console.log(`  Semifinal 2 (Midwest vs West):`);
console.log(`    ${semi2.winner.padEnd(20)} def. ${semi2.loser.padEnd(20)} (${(semi2.prob*100).toFixed(0)}% | margin: ${Math.abs(semi2.margin).toFixed(1)})`);

// ── Championship ─────────────────────────────────────────────────────────────
console.log(`\n${'━'.repeat(22)} NATIONAL CHAMPIONSHIP ${'━'.repeat(22)}`);
console.log(`  April 6, 2026 — Lucas Oil Stadium, Indianapolis\n`);

const champ = simGame(semi1.winner, semi2.winner);
console.log(`    ${champ.winner.padEnd(20)} def. ${champ.loser.padEnd(20)} (${(champ.prob*100).toFixed(0)}% | margin: ${Math.abs(champ.margin).toFixed(1)})`);

console.log(`\n${'═'.repeat(60)}`);
console.log(`  NATIONAL CHAMPION: ${champ.winner}`);
console.log(`${'═'.repeat(60)}`);

// ── Summary ──────────────────────────────────────────────────────────────────
console.log('\n BRACKET SUMMARY');
console.log('─'.repeat(40));
const regionNames = Object.keys(regions);
console.log(`  Final Four:`);
for (let i = 0; i < finalFourTeams.length; i++) {
  console.log(`    ${regionNames[i].padEnd(8)}: (${getSeed(regionNames[i], finalFourTeams[i])}) ${finalFourTeams[i]}`);
}
console.log(`  Champion:  ${champ.winner}`);
console.log(`  Runner-up: ${champ.loser}\n`);

// ── Upset alerts ─────────────────────────────────────────────────────────────
console.log('  NOTABLE UPSETS TO WATCH:');
const allGames = [];
function trackGame(g, round, seedW, seedL) {
  if (seedW > seedL && seedW - seedL >= 3) {
    allGames.push({ ...g, round, seedW, seedL });
  }
}
// Re-sim to collect upsets (we already printed results above)
// This is just for the summary
console.log('  (Teams where a lower seed wins by 3+ seed lines)');
console.log('─'.repeat(40));
